twilight-model/src/application/interaction/application_command/option.rs

```diff
@@ -83,17 +86,20 @@ impl<'de> Deserialize<'de> for CommandDataOption {
             Focused,
         }
 
-        // Id before string such that IDs will always be interpreted
-        // as such, this does mean that string inputs that looks like
-        // IDs will have to be caught if it is a string.
+        // An `Id` variant is purposely placed below `String` here.
+        // We first deserialize into a `String` and then
+        // later parse it into an `Id` variant. This is done
+        // to prevent situations where an option's string value
+        // is parsed as an `Id` variant, which omits features leading zeros.
+        // The `Id` variant is only used when the type is an `Id`.
         #[derive(Debug, Deserialize)]
         #[serde(untagged)]
         enum ValueEnvelope {
             Boolean(bool),
             Integer(i64),
             Number(f64),
-            Id(Id<GenericMarker>),
             String(String),
+            Id(Id<GenericMarker>),
```

Comment:
As it's, from my understanding, impossible to deserialize into this variant, might it not make more sense to remove it outright?

If we remove it you can also revert the `match` blocks back to `if let ... else` blocks

Reply:
I originally thought this as well, my prior commits omitted the variant however it seemed to cause the tests to fail because it couldn't figure out how to deserialize `Id`'s with `ValueEnvelope`. I had assumed it would be always ingested as a string, so I'm unsure why it fails there.

Reference to failing test:
- https://github.com/twilight-rs/twilight/actions/runs/4598981465/jobs/8123838189

Reply:
Weird, I'm unable to reproduce this locally. Even on your exact commit lol.

Reply:
Yeah no idea why that is. Intresstingly, removing the `ValueEnvelope::String` arms don't cause any errors (unreachable?) so to provide code clarity down the line I suggest we remove / panic on them.

Reply:
> I originally thought this as well, my prior commits omitted the variant however it seemed to cause the tests to fail because it couldn't figure out how to deserialize `Id`'s with `ValueEnvelope`. I had assumed it would be always ingested as a string, so I'm unsure why it fails there.
> 
> Reference to failing test:
> 
> * https://github.com/twilight-rs/twilight/actions/runs/4598981465/jobs/8123838189

The issue is that this also tests deserialization. So if `assert_ser_tokens` would be used instead the tests would pass. Idk how to fix this error rn would need to look more into it. So as a temp fix I would say we should switch to `assert_ser_tokens` and not use `assert_tokens`.
Keeping `ValueEnvelope::Id` cuz of a broken test isn't sth. we should do imo.